test/distributed/test_c10d.py

```diff
@@ -3825,6 +3827,51 @@ def test_broadcast_coalesced_gloo_cpu(self):
             self._test_broadcast_coalesced(process_group, device, root_rank)
 
 
+class TestDistributedNNFunctions(MultiProcessTestCase):
+    def setUp(self):
+        super(TestDistributedNNFunctions, self).setUp()
+        os.environ['WORLD_SIZE'] = str(self.world_size)
+        os.environ['MASTER_ADDR'] = '127.0.0.1'
+        os.environ['MASTER_PORT'] = "12345"
+        # This doesnt work ...
+        os.environ['MASTER_PORT'] = str(common.find_free_port())
+        self._fork_processes()
+        os.environ['RANK'] = str(self.rank)
+
+    def tearDown(self):
+        super(TestDistributedNNFunctions, self).tearDown()
+        try:
+            os.remove(self.file_name)
+        except OSError:
+            pass
+
+    @property
+    def op_timeout_sec(self):
+        return 1
+
+    @property
+    def world_size(self):
+        return 2
+
+    @requires_nccl()
+    @skip_if_not_multigpu
+    def test_broadcast(self):
+        dist.init_process_group(init_method='env://', rank=self.rank, world_size=self.world_size, backend='nccl')
+        device = torch.device("cuda:%d" % self.rank)
+        x = torch.zeros(5, 5).to(device) + self.rank
+        x.requires_grad = True
+        y = torch.distributed.nn.broadcast(x, 1)
+        self.assertEqual(y, torch.ones(5, 5))
+        z = torch.nn.MSELoss()(x, y)
+        # z.backward()
+        def fn(x):
+            y = torch.distributed.nn.broadcast(x, 1)
+            y = torch.sin(y)
+            return torch.nn.MSELoss()(x, y)
+
+        torch.autograd.gradcheck(fn, (x.clone().requires_grad_(),))
```

Comment:
I did a test with this  (I verified that the gradients match with chainermn)
but since communication collectives are not differentiable, I think that doing 
`gradcheck` is not possible. BTW it errors in this case.

I think that we will need to manually calculate the gradient of the test case by ourselves.

Reply:
`gradcheck` will call the function's forward a few times to compute the gradient with finite difference. Then backward a few times to compute the gradients using autograd and compare the two.
But I see how this might not work with distributed primitives :D 

But I guess you can design a small test for which the gradients are simple and check that the communications happen properly and you get the right result?